protoc-gen-openapiv2/internal/genopenapi/template.go

```diff
@@ -1004,14 +1015,72 @@ pathLoop:
 			}
 			// Pop from the stack
 			depth--
-			buffer += string(char)
-			if reg.GetUseJSONNamesForFields() &&
-				len(jsonBuffer) > 1 {
-				jsonSnakeCaseName := jsonBuffer[1:]
-				jsonCamelCaseName := lowerCamelCase(jsonSnakeCaseName, fields, msgs)
-				prev := buffer[:len(buffer)-len(jsonSnakeCaseName)-2]
-				buffer = strings.Join([]string{prev, "{", jsonCamelCaseName, "}"}, "")
-				jsonBuffer = ""
+			paramName := ""
+			pattern := ""
+			if reg.GetExpandSlashedPathPatterns() {
+				paramPattern := strings.SplitN(buffer[1:], "=", 2)
+				if len(paramPattern) == 2 {
+					paramName = paramPattern[0]
+					pattern = paramPattern[1]
+				}
+			}
+			if pattern == "" || pattern == "*" {
+				buffer += string(char)
+				if reg.GetUseJSONNamesForFields() &&
+					len(jsonBuffer) > 1 {
+					jsonSnakeCaseName := jsonBuffer[1:]
+					jsonCamelCaseName := lowerCamelCase(jsonSnakeCaseName, fields, msgs)
+					prev := buffer[:len(buffer)-len(jsonSnakeCaseName)-2]
+					buffer = strings.Join([]string{prev, "{", jsonCamelCaseName, "}"}, "")
+					jsonBuffer = ""
+				}
+				continue
+			}
+
+			pathParamIndex := slices.IndexFunc(*pathParams, func(p descriptor.Parameter) bool {
+				return p.FieldPath.String() == paramName
+			})
+			if pathParamIndex == -1 {
+				panic(fmt.Sprintf("Path parameter %q not found in path parameters", paramName))
+			}
+			pathParam := (*pathParams)[pathParamIndex]
+			patternParts := strings.Split(pattern, "/")
+			for i, part := range patternParts {
+				modifiedPart := part
+				if part == "*" {
+					paramName := strings.TrimSuffix(parts[len(parts)-1], "s")
+					if reg.GetUseJSONNamesForFields() {
+						paramName = casing.JSONCamelCase(paramName)
+					}
+					modifiedPart = "{" + paramName + "}"
+					newParam := descriptor.Parameter{
+						Target: &descriptor.Field{
+							FieldDescriptorProto: &descriptorpb.FieldDescriptorProto{
+								Name: proto.String(paramName),
+							},
+							Message:           pathParam.Target.Message,
+							FieldMessage:      pathParam.Target.FieldMessage,
+							ForcePrefixedName: pathParam.Target.ForcePrefixedName,
+						},
+						FieldPath: []descriptor.FieldPathComponent{{
+							Name:   paramName,
+							Target: nil,
+						}},
+						Method: nil,
+					}
+					*pathParams = append((*pathParams), newParam)
+					if pathParamIndex != -1 {
+						// the new parameter from the pattern replaces the old path parameter
+						*pathParams = append((*pathParams)[:pathParamIndex], (*pathParams)[pathParamIndex+1:]...)
+						pathParamIndex = -1
+					}
+				}
+				isLastPart := i == len(patternParts)-1
+				if isLastPart {
+					buffer = modifiedPart
+				} else {
+					parts = append(parts, modifiedPart)
+				}
```

Comment:
This is a lot of logic to add to an already complicated function. Do you think you could break this out into a separate function with separate tests, so we can cover the different scenarios more easily?